src/lsm/segmented_array.zig

```diff
@@ -219,98 +284,82 @@ fn SegmentedArrayType(
-            assert(total == a_half + b_half);
-
-            if (existing_a_tail.len > 0) assert(existing_b_head.len == 0 and elements_b.len == 0);
-            if (existing_b_head.len > 0) assert(elements_a.len == 0 and existing_a_tail.len == 0);
-            if (elements_a.len > 0) assert(existing_b_head.len == 0);
-            if (elements_b.len > 0) assert(existing_a_tail.len == 0);
-
-            assert(existing_a_head.ptr == a_pointer);
-            assert(existing_a_head.ptr + existing_a_head.len == existing_a_tail.ptr);
-            assert(existing_a_head.ptr + existing_a_head.len == existing_b_head.ptr);
-            if (existing_b_head.len > 0) {
-                assert(existing_b_head.ptr + existing_b_head.len == existing_b_tail.ptr);
-            } else {
-                assert(existing_a_tail.ptr + existing_a_tail.len == existing_b_tail.ptr);
-            }
+            const a_half_pointer = a_pointer[0..a_half];
+            const b_half_pointer = b_pointer[0..b_half];
 
-            mem.copy(T, b_pointer[existing_b_head.len + elements_b.len ..], existing_b_tail);
-            mem.copy(T, b_pointer[existing_b_head.len..], elements_b);
-            mem.copy(T, b_pointer, existing_b_head);
+            // Move part of `a` forwards to make space for elements.
+            copy_backwards(
+                a_half_pointer,
+                b_half_pointer,
+                cursor.relative_index + elements.len,
+                a_pointer[cursor.relative_index..a_count],
+            );
 
-            mem.copyBackwards(
-                T,
-                a_pointer[existing_a_head.len + elements_a.len ..],
-                existing_a_tail,
+            if (a_half < cursor.relative_index) {
+                // Move the part of `a` that is past the half-way point into `b`.
+                mem.copyBackwards(
+                    T,
+                    b_half_pointer,
+                    a_pointer[a_half..cursor.relative_index],
+                );
+            }
+
+            // Move `elements` into `a` and/or `b`.
+            copy_backwards(
+                a_half_pointer,
+                b_half_pointer,
+                cursor.relative_index,
+                elements,
             );
-            mem.copy(T, a_pointer[existing_a_head.len..], elements_a);
 
             array.indexes[b] = array.indexes[a] + a_half;
             array.increment_indexes_after(b, @intCast(u32, elements.len));
         }
 
+        /// Behaves like mem.copyBackwards, but as if `a` and `b` were a single contiguous slice.
+        fn copy_backwards(
+            a: []T,
+            b: []T,
+            dest: usize,
+            source: []const T,
+        ) void {
+            assert(dest + source.len <= a.len + b.len);
+            const dest_a = a[@minimum(dest, a.len)..@minimum(dest + source.len, a.len)];
+            const dest_b = b[dest -| a.len..(dest + source.len) -| a.len];
```

Comment:
How about instead of `dest` we use `target`? That way it is a full word, and it even has the name number of letters as "source" so the arguments will align :grin:.